### ISSUE TITLE
nvim crashes by nvim_buf_set_lines() when undolevels = -1

### ISSUE THREAD

**Issue by wookayin** [neovim/neovim#24894]
### Problem

Nvim crashes when the number of lines gets larger than the window height -- but this is possibly due to a plugin. ~~The root cause might be on somewhere `extmark`; I will try to modify the title and the reproduce step if I can find why the crash happens.~~

I find that this happens when we call `nvim_buf_set_lines` for a buffer with `undolevels = -1`.

With git bisect I figured out that 008154954 (#24824) is the first bad commit. /cc @bfredl

Stacktraces tell that `u_force_get_undo_header` via `extmark_splice_impl` is where it aborts.

<details>
   <summary>Stacktrace</summary>

```
Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib        	       0x18b62e868 __pthread_kill + 8
1   libsystem_pthread.dylib       	       0x18b665cec pthread_kill + 288
2   libsystem_c.dylib             	       0x18b59e2c8 abort + 180
3   nvim                          	       0x102a42fa0 u_force_get_undo_header + 212 (undo.c:3220)
4   nvim                          	       0x10281f0b8 u_extmark_copy + 60 (extmark.c:455)
5   nvim                          	       0x10281f650 extmark_splice_impl + 292 (extmark.c:625)
6   nvim                          	       0x10282001c extmark_splice + 204 (extmark.c:603)
7   nvim                          	       0x1026ba524 nvim_buf_set_lines + 1716 (buffer.c:470)
8   nvim                          	       0x1026a03ec nlua_api_nvim_buf_set_lines + 524 (lua_api_c_bindings.generated.c:575)
9   nvim                          	       0x102b0d074 lj_BC_FUNCC + 44
10  nvim                          	       0x102b22f78 lua_pcall + 228 (lj_api.c:1145)
11  nvim                          	       0x10287abf8 nlua_schedule_event + 80 (executor.c:360)
12  nvim                          	       0x1029e9ed8 state_handle_k_event + 88 (state.c:117)
13  nvim                          	       0x10276f568 insert_handle_key + 2816 (edit.c:880)
14  nvim                          	       0x102768ac8 insert_execute + 1812 (edit.c:671)
15  nvim                          	       0x1029e9e44 state_enter + 600 (state.c:99)
16  nvim                          	       0x102769908 insert_enter + 2280 (edit.c:338)
17  nvim                          	       0x102768358 edit + 432 (edit.c:1266)
18  nvim                          	       0x1028f201c invoke_edit + 132 (normal.c:6270)
19  nvim                          	       0x1028ec8b0 nv_edit + 680 (normal.c:6247)
20  nvim                          	       0x1028e9aac normal_execute + 2144 (normal.c:1226)
21  nvim                          	       0x1029e9e44 state_enter + 600 (state.c:99)
22  nvim                          	       0x1028e5640 normal_enter + 196 (normal.c:517)
23  nvim                          	       0x10269733c main + 4292 (main.c:647)
24  dyld                          	       0x18b33be50 start + 2544
```
</details>

### Steps to reproduce

`$ nvim -u minimal.lua` where minimal.lua:

```
local buf = vim.api.nvim_create_buf(false, true)
vim.api.nvim_buf_set_option(buf, 'undolevels', -1)
vim.api.nvim_buf_set_lines(buf, 0, 1, false, { })
```

### Steps to reproduce (old)

<details>
   <summary>Click to expand</summary>

This happens when I use [nvim-scrollview](https://github.com/dstein64/nvim-scrollview) written by @dstein64, so I will provide steps with the plugin enabled:

`$ nvim -u minimal.lua` where minimal.lua:

```
local root = vim.fn.fnamemodify("./.repro", ":p")
for _, name in ipairs({ "config", "data", "state", "cache" }) do
  vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({ "git", "clone", "--filter=blob:none", "https://github.com/folke/lazy.nvim.git", lazypath, })
end
vim.opt.runtimepath:prepend(lazypath)
local plugins = {
  'dstein64/nvim-scrollview';
}
require("lazy").setup(plugins, {
  root = root .. "/plugins",
})
```

And then

- Enter `<CR>` sufficient times to make scrollbar appear (e.g., the number of lines in the buffer exceeds the window height)
- or open any files with many enough lines

~I will try to narrow down to the root cause with (more) minimal steps to reproduce.~ UPDATE: See above.

</details>

### Expected behavior

No crashes.

### Neovim version (nvim -v)

0.10.0-dev 008154954

### Vim (not Nvim) behaves the same?

N/A

### Operating system/version

macOS 12.3

### Terminal name/version

xterm

### $TERM environment variable

xterm-256color

### Installation

build from source (HEAD)

**Comment by wookayin** [comment]
Reading #24824, it's going to fixed by #24889. Sorry for a dup, but I will have this issue open remaining

**Comment by bfredl** [comment]
should be fixed by updated #24889 . I'll include your repro as a test, thanks!

**Comment by gegoune** [comment]
@bfredl I can confirm #24889 fixes this issue.
(it's not suppose to fix other, Telescope issue, is it?)
